transports/webrtc/src/connection/poll_data_channel.rs

```diff
@@ -0,0 +1,289 @@
+    Reading(Pin<Box<dyn Future<Output = Result<Vec<u8>, Error>> + Send + 'a>>),
+    /// Finished reading, but there's unread data in the temporary buffer.
+    RemainingData(Vec<u8>),
+}
+
+impl<'a> ReadFut<'a> {
+    /// Gets a mutable reference to the future stored inside `Reading(future)`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `ReadFut` variant is not `Reading`.
+    fn get_reading_mut(
+        &mut self,
+    ) -> &mut Pin<Box<dyn Future<Output = Result<Vec<u8>, Error>> + Send + 'a>> {
+        match self {
+            ReadFut::Reading(ref mut fut) => fut,
+            _ => panic!("expected ReadFut to be Reading"),
+        }
+    }
+}
+
+/// A wrapper around around [`DataChannel`], which implements [`AsyncRead`] and
+/// [`AsyncWrite`].
+///
+/// Both `poll_read` and `poll_write` calls allocate temporary buffers, which results in an
+/// additional overhead.
+pub struct PollDataChannel<'a> {
+    data_channel: Arc<DataChannel>,
+
+    read_fut: ReadFut<'a>,
+    write_fut: Option<Pin<Box<dyn Future<Output = Result<usize, Error>> + Send + 'a>>>,
+    shutdown_fut: Option<Pin<Box<dyn Future<Output = Result<(), Error>> + Send + 'a>>>,
+
+    read_buf_cap: usize,
+}
+
+impl PollDataChannel<'_> {
+    /// Constructs a new `PollDataChannel`.
+    pub fn new(data_channel: Arc<DataChannel>) -> Self {
+        Self {
+            data_channel,
+            read_fut: ReadFut::Idle,
+            write_fut: None,
+            shutdown_fut: None,
+            read_buf_cap: DEFAULT_READ_BUF_SIZE,
+        }
+    }
+
+    /// Get back the inner data_channel.
+    pub fn into_inner(self) -> Arc<DataChannel> {
+        self.data_channel
+    }
+
+    /// Obtain a clone of the inner data_channel.
+    pub fn clone_inner(&self) -> Arc<DataChannel> {
+        self.data_channel.clone()
+    }
+
+    /// Set the capacity of the temporary read buffer (default: 4096).
+    pub fn set_read_buf_capacity(&mut self, capacity: usize) {
+        self.read_buf_cap = capacity
+    }
+}
+
+impl AsyncRead for PollDataChannel<'_> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let fut = match self.read_fut {
+            ReadFut::Idle => {
+                // read into a temporary buffer because `buf` has an unonymous lifetime, which can
+                // be shorter than the lifetime of `read_fut`.
+                let dc = self.data_channel.clone();
+                let mut temp_buf = vec![0; self.read_buf_cap];
+                self.read_fut = ReadFut::Reading(Box::pin(async move {
+                    let res = dc.read(temp_buf.as_mut_slice()).await;
+                    match res {
+                        Ok(n) => {
+                            temp_buf.truncate(n);
+                            Ok(temp_buf)
+                        },
+                        Err(e) => Err(e),
+                    }
+                }));
```

Comment:
This uses `async move`, so the lifetime seems to _effectively_ always be `'static` so we shouldn't need to have `'a` within here I believe.